### ISSUE TITLE
Remove redundant exit checks in lab_to_target

### ISSUE THREAD

**Issue by tanyongkiam** [CakeML/cakeml#107]
The checks are:
done /\ all_enc_ok c enc labs 0 sec_list /\ labs2 = labs /\
             EVERY (check_lab sec_list) (all_labels labs) /\
             ALL_DISTINCT (sec_names sec_list)

The important one (i.e. the slowest one) to remove is: EVERY (check_lab sec_list) (all_labels labs).

**Comment by tanyongkiam** [comment]
This was partially done in 99890f3

Remaining checks are more difficult (I think):
done /\ all_enc_ok c enc labs 0 sec_list /\ labs2 = labs

**Comment by tanyongkiam** [comment]
all_enc_ok should probably be done after #109

**Comment by xrchz** [comment]
`all_enc_ok` is extremely slow for bootstrapping (because there are so many lines: each one takes <1s, but they add up), so I would treat it as highest priority to remove. The other checks are not so bad to bootstrap, but removing more would be better.

**Comment by tanyongkiam** [comment]
From a quick look, I think enc_with_nop is relatively easy to remove (and it is probably a big culprit in the slowness)

I don't know why checking each line takes 1s though, that seems way too high, especially since all the encodings should be cached already.

**Comment by myreen** [comment]
Note that when we talk about removing the a condition from `all_enc_ok`. The right thing is to:
 - copy `all_enc_ok` and its `line_ok` helper function in `lab_to_targetProof`, and
 - write a simpler version `check_encs` (or some similar name) that the compiler can use where it used to have `all_enc_ok`, then
 - prove that the context where `check_encs` appears in the assembler implies that the original `all_enc_ok` is true 

I agree with @tanyongkiam that `enc_with_nop` should be fairly easy to remove. The length checks, i.e. `LENGTH bytes = l`, should also be easy enough to remove.

With some effort, it is probably possible to remove `asm_ok` from the `Asm` case in `line_ok`. In order to remove this `asm_ok`, you'll have to propagate this syntactic condition up through the compiler phases. There is already a placeholder for such syntactic conditions: `lab_to_targetProof` defines:

    good_syntax mc_conf (code:'a sec list) (l:num_set) = T

This should be changed to say that all `Asm` lines contain something that is `asm_ok`. This will mostly be a mechanical refactoring to bubble this up. However, there are instructions in `stack_remove` which might not be `asm_ok`. Here the compiler ought to be updated to ensure that all stack accessing loads and stores are `asm_ok`. One can do this by replacing an impossible `load r, mem[sp+40000]` with code that uses smaller immediate constants. This example is assuming that  `40000` is too big as an immediate offset here. For example, one could generate the following in these kinds of pathological cases. These pathological cases arise when a stack frame is *way* too big. 

    add sp,sp,10000
    add sp,sp,10000
    add sp,sp,10000
    add sp,sp,10000 
    load r, mem[sp]
    sub sp,sp,10000
    sub sp,sp,10000
    sub sp,sp,10000
    sub sp,sp,10000 

This would be a nice change in that this makes the compiler more total.

**Comment by xrchz** [comment]
Progress made in 2f70c1cf2a9834582ea8c4f260aa55d6420164ed

**Comment by xrchz** [comment]
The `all_enc_ok` check is still way too slow - at a rough calculation, it looks like it would take about 1 year of CPU time. I hope @tanyongkiam or @myreen or someone else can look into getting rid of the `asm_ok` checks. Then, I hope we can follow the suggestion of @SOwens to omit the jump target checks (@myreen, this would mean revisiting your proofs). That would remove `all_enc_ok` completely. (Meanwhile you may have seen I removed the redundant "compute labels again"; that shaves off about half an hour.)

**Comment by myreen** [comment]
I can look at removing the `lab_lookup l1 l2 labs` check as suggested by @SOwens. 

Regarding `asm_ok`, I believe it can be removed from the `Asm b bytes l` case, but not from the other cases because the compiler can, for valid reasons, fail to encode a LabLang program at this point. Consider for example encoding `Loc` for a label that is 4000 bytes away. For x86-64, I believe encoding a 4000 byte offset is no problem, while on other architectures this might not be possible with the current instruction encodings. Even if we make it accept larger offsets, one can still hit this issue. I believe x86-64 struggles if the offset is more than 4 gigabytes away.

@xrchz Could you check what kind of speed up one gets if the `lab_lookup` is removed? Similarly, is there a noticeable speed up if `asm_ok` is removed only for the `Asm b bytes l` case?

**Comment by myreen** [comment]
I noticed that the current definition of `line_ok_light` for case `LabAsm a w bytes l` does a lookup twice on each label: once in `get_label` and another time in `lab_lookup`. It should be very easy to rephrase this to only do one lookup. I'll do that.

However, an even better improvement would be to remove all of the `let w1 = ... in` bindings from `line_ok` and use `w` from the input. The value of `w1` and `w` ought to be the same.

@xrchz could you check what the speed of evaluation would be if `line_ok_light` was defined as follows?

    (line_ok_light (c:'a asm_config) labs pos (Label _ _ l) <=> T) /\
    (line_ok_light c labs pos (Asm b bytes l) <=> asm_ok b c) /\
    (line_ok_light c labs pos (LabAsm Halt w bytes l) <=>
       asm_ok (Jump w) c) /\
    (line_ok_light c labs pos (LabAsm ClearCache w bytes l) <=>
       asm_ok (Jump w) c) /\
    (line_ok_light c labs pos (LabAsm (CallFFI index) w bytes l) <=>
       asm_ok (Jump w) c) /\
    (line_ok_light c labs pos (LabAsm (Call v24) w bytes l) <=>
       F (* Call not yet supported *)) /\
    (line_ok_light c labs pos (LabAsm a w bytes l) <=>
       asm_ok (lab_inst w a) c)

**Comment by xrchz** [comment]
I can't easily check, but based on what I have seen and know, I would say:
  - Removing `asm_ok` just for the `Asm` case would be a big improvement, because the majority of instructions are `Asm`. Computing `asm_ok` for a handful of `LabAsm` instructions shouldn't be too bad.
  - Removing `labs` from `line_ok_light` (your revised definition does not need it) would be a very big improvement, because a lot of the slowdown seems to come from expanding out the definition of `labs` because it is a huge term.
I think trying to prove that revised version of `line_ok_light` sufficient is worthwhile. I will try it.

**Comment by myreen** [comment]
My revised definition requires digging into my semantic proofs for this phase (because label lookups might fail now). If you can cheat the places where those semantic proofs break, then I can look at them on the weekend (very busy until then).

**Comment by myreen** [comment]
I made some progress on removing `lab_lookup` in bd12fee. The commit message explains that many semantics (BVI, dataLang, wordLang, stackLang, labLang) need very minor tweaks.

**Comment by tanyongkiam** [comment]
@xrchz has made a lot of progress on this.

New issue recently: we ran into a legitimate immediate jump that was too big for the encoding in unoptimized MIPS.

I am removing part of the `asm_ok` check on the asm_ok branch

**Comment by xrchz** [comment]
what is left now?

**Comment by tanyongkiam** [comment]
This asm_ok check could be simplified in a few of the cases (JumpCmp, reg_ok):
https://github.com/CakeML/cakeml/blob/master/compiler/backend/lab_to_targetScript.sml#L191

The other checks depend on the assembler.

**Comment by xrchz** [comment]
I think we could close this issue now. The last few checks probably don't make a big difference. Maybe they could be part of a separate issue. What do you think? Magnus suggested that caching the evaluation of something inside `asm_ok` could get most of the benefit.
